CONTRIBUTING.md

```diff
@@ -0,0 +1,155 @@
+
+- [`ovl_Dm_Ravine`](src/overlays/actors/ovl_Dm_Ravine/z_dm_ravine.c)
+- [`ovl_En_Jc_Mato`](src/overlays/actors/ovl_En_Jc_Mato/z_en_jc_mato.c)
+
+These files demonstrate the following:
+
+- Word order in names are from least-to-most specific (`DM_RAVINE_STATE_ACTIVE`, not `DM_RAVINE_ACTIVE_STATE`)
+- Functions, structs, unions, enums, and typedefs are `TitleCase` (`DmRavine`)
+    - "Methods" for objects separate the object from the verb with an underscore (`DmRavine_Init`)
+- Variable names are `camelCase` (`actionFunc`)
+    - Global variables start with `g` (`gSaveContext`)
+    - Static global variables start with `s` (`sSphereInit`)
+- Macros and enum constants are `SCREAMING_SNAKE_CASE` (`DM_RAVINE_STATE_ACTIVE`)
+- Public structs, enums, and constants should be defined in the header file in the following cases:
+    - The primary `Actor` struct (`DmRavine`)
+    - The actor's `InitVars`, as an `extern` symbol (`Dm_Ravine_InitVars`)
+    - Types used in other public declarations, like `actionFunc` typedefs (`DmRavineActionFunc`)
+    - Macros and enums for accessing & setting actor `params`
+    - The struct/enum/define is needed by other files
+- Trailing commas in array and struct definitions (see `EnJcMatoDamageTable`)
+- Constants converted to whichever looks best in context: hexadecimal, decimal, or float
+    - Rotation angles should always be in hexadecimal
+- Define macros for bitwise access to `actor->params`
+
+### `NON_MATCHING` & `NON_EQUIVALENT`
+
+If you make significant progress on decompiling a function, but can't get it to match perfectly, you can use a `NON_MATCHING` block to commit your code but have it excluded from the build, like so:
+
+```c
+#ifdef NON_MATCHING
+void CollisionCheck_SpawnWaterDroplets(GlobalContext* ctxt, Vec3f* v) {
+    // ... 
+}
+#else
+#pragma GLOBAL_ASM("./asm/non_matchings/code/z_collision_check/CollisionCheck_SpawnWaterDroplets.asm")
+#endif
+```
+
+Before using `NON_MATCHING`, first try to use the [decomp-permuter](tools/decomp-permuter) tool to find a closer match.
+
+`NON_EQUIVALENT` can be used with the same syntax as `NON_MATCHING`, but it is used to mark sections of code which do not match *and* do not have the same behavior as the original code.
+
+### Matching vs. Documenting
+
+Usually, the first step of decompiling a section of code is to get it *matching*: to produce a C version of the code that can be compiled into an identical ROM.
+
+However, the goal of this project is to produce a codebase that can be understood and modified.
+So, beyond producing matching C code, the next steps are *documenting* the code.
+
+Documenting is more than just adding comments. Documenting also includes:
+
+- Renaming functions, variables, and struct members
+- Using (or adding) constants, enums, and macros when possible
+- Explaining sections of code that are not straightforward
+
+Overlays are not required to be documented at this time, but `code/` and `boot/` are.
```

Comment:
I don't really agree that anything should be "required" to be documented. Just having decompiled code is a huge step forward and requiring people who might not have the time to sit down and fully understand the system they are decompiling at the time seems inappropriate. Not to mention, certain core systems that operate in close connection with other systems or make use of subsystems may be unreasonable to document properly at the time and will need to be revisited after all other relevant information is available to ensure the documentation is accurate.

Reply:
Good point that there are exceptions when one system calls on another system a lot. There should probably be a carve out exception for those situations. That being said, in most cases that documentation these days can be gleaned from looking at OOT code.

As far as someone not having time, to be honest, my solution to them would be to work on an actor where documentation is not required.

Reply:
Yeah this makes sense. I think avoiding saying "X is required" is a net positive, it sounds a bit rigid.

```suggestion
Overlays are not required to be documented at this time, but `code/` and `boot/` should be documented as much as possible.
```

@Kenix3 what do you think? Do you have any suggestions here?

Reply:
I'm happy with that rewording.